### PR TITLE
Add basic tensorboard support

### PR DESCRIPTION
**Pull request by zcain117**
Tested on a fresh red VM that I built from head using build_torch_wheels.sh

### CODE REVIEW

**Comment by zcain117** [comment]
@jysohn23 how would I add `tensorboard` to the pytorch-nightly and pytorch-0.1 conda environments that come with the pytorch image? And also to the docker images?

File: test/test_train_imagenet.py [hunk 1]
```diff
@@ -164,10 +166,13 @@ def test_loop_fn(model, loader, device, context):
...
   accuracy = 0.0
+  writer = SummaryWriter(log_dir='/tmp/imagenet_tensorboard')
   for epoch in range(1, FLAGS.num_epochs + 1):
     model_parallel(train_loop_fn, train_loader)
     accuracies = model_parallel(test_loop_fn, test_loader)
-    accuracy = sum(accuracies) / len(accuracies)
+    print("Epoch: {}".format(epoch))
+    accuracy = mean(accuracies)
+    writer.add_scalar('Accuracy/test', accuracy, epoch)
     if FLAGS.metrics_debug:
```
Comment (dlibenzi) [inline test/test_train_imagenet.py]: We should probably print the accuracy mean as well.

File: test/test_train_imagenet.py [hunk 2]
```diff
@@ -164,10 +166,13 @@ def test_loop_fn(model, loader, device, context):
     return correct / total_samples
 
   accuracy = 0.0
+  writer = SummaryWriter(log_dir='/tmp/imagenet_tensorboard')
```
Comment (jysohn23) [inline test/test_train_imagenet.py]: Could we add FLAGS.logdir? What do you think?

Comment (dlibenzi) [inline test/test_train_imagenet.py]: This should definitely be the default, and the `log_dir` must be a command line argument.

File: test/test_train_imagenet.py (hunk 1 above)
Comment (zcain117) [reply test/test_train_imagenet.py]: Done

File: test/test_train_imagenet.py [hunk 3]
```diff
@@ -164,10 +166,13 @@ def test_loop_fn(model, loader, device, context):
...
   for epoch in range(1, FLAGS.num_epochs + 1):
     model_parallel(train_loop_fn, train_loader)
     accuracies = model_parallel(test_loop_fn, test_loader)
-    accuracy = sum(accuracies) / len(accuracies)
+    print("Epoch: {}".format(epoch))
+    accuracy = mean(accuracies)
+    writer.add_scalar('Accuracy/test', accuracy, epoch)
```
Comment (jysohn23) [inline test/test_train_imagenet.py]: Can we add additional scalar metrics such as loss, examples/sec, etc (where the x-axis is global step number)? Similar to the tensorboard metrics we see with TF estimator. We can do this in a followup PR maybe?

Comment (zcain117) [reply test/test_train_imagenet.py]: I'll be adding more in some followup PRs. Still deciding on a clean way to do the others since we'll want to average over all devices during the training loop. Or we could have N different lines on the loss curves where N=num_cores

File: test/test_train_imagenet.py (hunk 2 above)
Comment (zcain117) [reply test/test_train_imagenet.py]: Defaulted to FLAGS.logdir

Comment (zcain117) [reply test/test_train_imagenet.py]: Defaulted to FLAGS.logdir

**Comment by zcain117** [comment]
> Can we cleanup unintentionally added runs/Sep03_23-44-20_zcain-highmem-96-build-from-head/events.out.tfevents.1567554260.zcain-highmem-96-build-from-head.54803.0 etc output from tb? Thanks.

Yup yup, removed

File: test/test_train_imagenet.py [hunk 4]
```diff
@@ -164,10 +166,13 @@ def test_loop_fn(model, loader, device, context):
     return correct / total_samples
 
   accuracy = 0.0
+  writer = SummaryWriter(log_dir=FLAGS.logdir)
```
Comment (dlibenzi) [inline test/test_train_imagenet.py]: This needs to be created only if logdir is not None ... unless SummaryWriter is a noop if logdir is None.

Comment (zcain117) [reply test/test_train_imagenet.py]: I tried running this without specifying --logdir and verified that None was being passed into SummaryWriter. When this happens, SummaryWriter works without complaining and defaults to creating a new directory: "runs/"

Comment (dlibenzi) [reply test/test_train_imagenet.py]: No, we do not want that :wink: 
So lets create a write only if the user wants to (logdir not None).

Comment (zcain117) [reply test/test_train_imagenet.py]: If writer is None when logdir is None, we will need to perform a check for `if writer:` at every spot where we call `writer.add_scalar`. Not a huge deal but adds to clutter

Comment (dlibenzi) [reply test/test_train_imagenet.py]: That's OK. Dumping stuff into a random location is worse :wink:

Comment (zcain117) [reply test/test_train_imagenet.py]: Added a helper method so it'll be a 1-line call whenever we write a summary data point. Verified that nothing gets written with `--logdir=None` and that valid summaries get written with `--logdir=non-None`

File: test/test_train_imagenet.py [hunk 5]
```diff
@@ -73,6 +75,11 @@ def get_model_property(key):
   return MODEL_PROPERTIES.get(FLAGS.model, MODEL_PROPERTIES['DEFAULT'])[key]
 
 
+def add_scalar_to_summary(summary_writer, metric_name, y_value, x_value):
```
Comment (dlibenzi) [inline test/test_train_imagenet.py]: Move this to `test_utils.py`.
Also, what are those x,y values?

Comment (zcain117) [reply test/test_train_imagenet.py]: Moved and renamed x,y to more closely match the documentation: https://pytorch.org/docs/stable/tensorboard.html#torch.utils.tensorboard.writer.SummaryWriter.add_scalar

In my mind, each scalar is an x,y coordinate in the Tensorboard graph. But hopefully this new naming is more clear

**Comment by dlibenzi** [comment]
We may want to change MNIST and CIFAR10 as well...

**Comment by zcain117** [comment]
Verified that my CL on the google3 side has propagated and that pytorch-nightly conda env now has tensorboard installed

**Comment by zcain117** [comment]
This change works on pytorch-nightly but on pytorch-0.1, I needed to run `conda install future`. I can either add `future` to our conda setup file or just wait to see if it's needed once we cut a new stable build (which I think will be soon). @dlibenzi what do you think?

**Comment by dlibenzi** [comment]
Hmm, pytorch-0.1 should not be running newer code.
We are not forward compatible.

**Comment by zcain117** [comment]
> Hmm, pytorch-0.1 should not be running newer code.
> We are not forward compatible.

Ok sounds good. I will leave it alone for now and then we can add `conda install future` if necessary once we are cutting the new build
